### PR TITLE
[Merged by Bors] - [Merged by Bors] - feat(CategoryTheory): covariant functoriality of graded objects on the index set

### PR DESCRIPTION
**Pull request by joelriou**
---

This is part of #7389 which intends to define a monoidal category structure on the graded objects (by the natural numbers) of a monoidal category.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/GradedObject.lean [hunk 1]
```diff
@@ -80,6 +85,39 @@ def eval {β : Type w} (b : β) : GradedObject β C ⥤ C where
...
+/-- Constructor for isomorphisms in `GradedObject` -/
+@[simps]
+def isoMk (e : ∀ i, X i ≅ Y i) : X ≅ Y where
+  hom i := (e i).hom
+  inv i := (e i).inv
+
+variable {X Y}
+
+lemma isIso_of_isIso_apply (f : X ⟶ Y) [hf : ∀ i, IsIso (f i)] :
+    IsIso f := by
```
Comment (eric-wieser) [inline Mathlib/CategoryTheory/GradedObject.lean:99]: Should this be an `instance`?

Comment (joelriou) [reply Mathlib/CategoryTheory/GradedObject.lean:100]: I am not sure how to accurately test that it does create a loop, but I do not think it should be an instance, similarly as `NatIso.isIso_of_isIso_app` is not an instance. I have added a comment.

File: Mathlib/CategoryTheory/GradedObject.lean [hunk 2]
```diff
@@ -238,4 +277,142 @@ end
...
 end GradedObject
 
+namespace GradedObject
+
+variable {I J : Type*} {C : Type*} [Category C]
+  (X Y Z : GradedObject I C) (φ : X ⟶ Y) (e : X ≅ Y) (ψ : Y ⟶ Z) (p : I → J)
+
+/-- If `X : GradedObject I C` and `p : I → J`, `X.mapObjFun p j` is the family of objects `X i`
+for `i : I` such that `p i = j`. -/
+abbrev mapObjFun (j : J) := (fun (i : (p ⁻¹' {j})) => X i)
```
Comment (eric-wieser) [inline Mathlib/CategoryTheory/GradedObject.lean:287]: Can you write an explicit type here?
```suggestion
abbrev mapObjFun (j : J) :  (p ⁻¹' {j}) → sorry := fun i => X i
```

File: Mathlib/CategoryTheory/GradedObject.lean [hunk 3]
```diff
@@ -238,4 +277,142 @@ end
...
+variable [X.HasMap p] [Y.HasMap p] [Z.HasMap p]
+
+/-- Given `X : GradedObject I C` and `p : I → J`, `X.mapObj p` is the graded object by `J`
+which in degree `j` consists of the coproduct of the `X i` such that `p i = j`. -/
+noncomputable def mapObj : GradedObject J C := fun j => ∐ (X.mapObjFun p j)
+
+/-- The canonical inclusion `X i ⟶ X.mapObj p j` when `i : I` and `j : J` are such
+that `p i = j`. -/
+noncomputable def ιMapObj (i : I) (j : J) (hij : p i = j) : X i ⟶ X.mapObj p j :=
+  Sigma.ι (fun (i' : (p ⁻¹' {j})) => X i') ⟨i, hij⟩
```
Comment (eric-wieser) [inline Mathlib/CategoryTheory/GradedObject.lean:304]: I assume this was intended?
```suggestion
  Sigma.ι (mapObjFun X) ⟨i, hij⟩
```

File: Mathlib/CategoryTheory/GradedObject.lean [hunk 4]
```diff
@@ -238,4 +277,142 @@ end
...
+
+/-- Given `X : GradedObject I C`, `p : I → J` and `j : J`,
+`CofanMapObjFun X p j` is the type `Cofan (X.mapObjFun p j)`. The point object of
+such colimits cofans are isomorphic to `X.mapObj p j`, see `CofanMapObjFun.iso`. -/
+abbrev CofanMapObjFun (j : J) := Cofan (X.mapObjFun p j)
```
Comment (eric-wieser) [inline Mathlib/CategoryTheory/GradedObject.lean:309]: Here too; is this maybe `Type _`?
```suggestion
abbrev CofanMapObjFun (j : J) : Type _ := Cofan (X.mapObjFun p j)
```

File: Mathlib/CategoryTheory/GradedObject.lean (hunk 3 above)
Comment (joelriou) [reply Mathlib/CategoryTheory/GradedObject.lean]: Actually, no... Thanks for the suggestions!

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**